### PR TITLE
En_Changer OK

### PR DESCRIPTION
**Pull request by EllipticEllipsis**
Could have done with the Lens of Truth for that Init and data. Rest is much simpler, and the devs politely documented most of it already in the osSyncPrintfs.

### CODE REVIEW

**Review by fig02 — APPROVED** [review]
review changes implemented

File: src/overlays/actors/ovl_En_Changer/z_en_changer.h [hunk 1]
```diff
@@ -3,12 +3,26 @@
...
+    /* 0x0154 */ EnBox* rightChest;
+    /* 0x0158 */ EnBox* finalChest;
+    /* 0x015C */ s16 leftChestNum;
+    /* 0x015E */ s16 rightChestNum;
+    /* 0x0160 */ s16 leftChestGetItemId;
+    /* 0x0162 */ s16 rightChestGetItemId;
+    /* 0x0164 */ s16 chestOpened;
+    /* 0x0166 */ s16 timer;
+    /* 0x0168 */ s16 roomChestsOpened;
+    /* 0x016A */ char unk_16A[0x2];
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Changer/z_en_changer.h:25]: probably padding
```suggestion
```

File: src/overlays/actors/ovl_En_Changer/z_en_changer.c [hunk 2]
```diff
@@ -20,15 +35,257 @@ const ActorInit En_Changer_InitVars = {
...
+static Vec3f sRightChestPos[] = {
+    { 0.0f, 0.0f, 0.0f },        { 140.0f, 20.0f, -245.0f },  { 140.0f, 20.0f, -685.0f },
+    { 140.0f, 20.0f, -1125.0f }, { 140.0f, 20.0f, -1565.0f }, { 140.0f, 20.0f, -2005.0f },
+};
+
+static s32 sLoserGetItemIds[] = {
+    GI_NONE, GI_RUPEE_GREEN_LOSE, GI_RUPEE_GREEN_LOSE, GI_RUPEE_BLUE_LOSE, GI_RUPEE_BLUE_LOSE, GI_RUPEE_RED_LOSE,
+};
+
+static s32 sEtcParams[] = { 0x0000, 0x0008, 0x0008, 0x0009, 0x0009, 0x000A };
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Changer/z_en_changer.c:53]: Values from the `ItemEtceteraType` enum should probably be used here, except the 0 which corresponds to the first room so I guess it's nothing.

I also suggest naming the array `sItemEtcTypes`

File: src/overlays/actors/ovl_En_Changer/z_en_changer.c [hunk 3]
```diff
@@ -20,15 +35,257 @@ const ActorInit En_Changer_InitVars = {
...
+        hpChestParams = sTreasureFlags[5] | hpChestParams;
+        this->finalChest = (EnBox*)Actor_SpawnAsChild(&globalCtx->actorCtx, &this->actor, globalCtx, ACTOR_EN_BOX,
+                                                      20.0f, 20.0f, -2500.0f, 0, 0x7FFF, 0, hpChestParams);
+        if (this->finalChest != NULL) {
+            if (this->roomChestsOpened) {
+                Flags_SetTreasure(globalCtx, hpChestParams & 0x1F);
+                Actor_Kill(&this->actor);
+                return;
+            } else {
+                hpParams = ((gSaveContext.itemGetInf[1] & 0x800) ? (0xB) : (0xC)) & 0xFF;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Changer/z_en_changer.c:107]: The 0xB and 0xC here should probably also be values from the `ItemEtceteraType` enum.

File: src/overlays/actors/ovl_En_Changer/z_en_changer.c [hunk 4]
```diff
@@ -20,15 +35,257 @@ const ActorInit En_Changer_InitVars = {
...
+    leftChestParams = (sLoserGetItemIds[globalCtx->roomCtx.curRoom.num] << 5) | 0x4000;
+    new_var = temp_v1_3;
+    this->leftChestNum = new_var;
+    this->leftChestGetItemId = sLoserGetItemIds[globalCtx->roomCtx.curRoom.num];
+    leftItemParams = sEtcParams[globalCtx->roomCtx.curRoom.num];
+    leftChestParams |= new_var;
+    rightChestParams = new_var | 0x4E21;
+    this->rightChestNum = new_var | 1;
+    this->rightChestGetItemId = GI_DOOR_KEY;
+    rightItemEtceteraOffset = 0xD;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Changer/z_en_changer.c:129]: Same for 0xD here

File: src/overlays/actors/ovl_En_Changer/z_en_changer.c [hunk 5]
```diff
@@ -20,15 +35,257 @@ const ActorInit En_Changer_InitVars = {
...
+    if (Math_Rand_ZeroFloat(1.99f) < 1.0f) {
+        rightChestParams = (sLoserGetItemIds[globalCtx->roomCtx.curRoom.num] << 5) | 0x4000;
+        this->rightChestNum = new_var;
+        this->rightChestGetItemId = sLoserGetItemIds[globalCtx->roomCtx.curRoom.num];
+        rightItemEtceteraOffset = sEtcParams[globalCtx->roomCtx.curRoom.num];
+        leftChestParams = new_var | 0x4E21;
+        rightChestParams |= new_var;
+        this->leftChestNum = temp_v1_3 | 1;
+        this->leftChestGetItemId = GI_DOOR_KEY;
+        leftItemParams = 0xD;
```
Comment (Roman971) [inline src/overlays/actors/ovl_En_Changer/z_en_changer.c:140]: And here

**Comment by fig02** [comment]
changes made

**Review by Roman971 — APPROVED** [review]

**Pull request merged**